### PR TITLE
Add a database connection pool for maintenance jobs

### PR DESCRIPTION
**Pull request by jgpruitt**
## Description

Adding a dedicated pool of database connections for use by maintenance jobs such as the telemetry and vacuum engines.

## Merge requirements

Please take into account the following non-code changes that you may need to make with your PR:

- [x] CHANGELOG entry for user-facing changes
- [x] Updated the relevant documentation

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -74,6 +74,7 @@ The following subsections cover all CLI flags which promscale supports. You can
 | db.connections.reader-pool.size               | integer  | 30% of possible connections db | Maximum size of the reader pool of database connections. This defaults to 30% of max_connections allowed by the database.                                                      |
 | db.connections.writer-pool.size               | integer  | 50% of possible connections db | Maximum size of the writer pool of database connections. This defaults to 50% of max_connections allowed by the database.                                                      |
 | db.connections.writer-pool.synchronous-commit | boolean  |             false              | Enable/disable synchronous_commit on database connections in the writer pool.                                                                                                  |
+| db.connections.maint-pool.size                | integer  |               5                | Maximum size of the maintenance pool of database connections. This defaults to 5                                                                                               |
```
Comment (alejandrodnm) [inline docs/configuration.md:77]: Should we specify that `vacuum.parallelism` takes the connections from this pool? Or some guidelines like it should be at least `vacuum.parallelism` + whatever else

File: pkg/pgclient/config.go [hunk 2]
```diff
@@ -154,32 +158,76 @@ func (cfg *Config) GetConnectionStr() string {
...
+	if readerPoolSize+writerPoolSize+maintPoolSize > dbMaxConns {
+		return 0, 0, 0,
+			fmt.Errorf("reader-pool (size=%d) + writer-pool (size=%d) + maint-pool (size=%d) more than postgres max_connections (%d). Decrease the pool-sizes", readerPoolSize, writerPoolSize, maintPoolSize, dbMaxConns)
 	}
-	if cfg.UsesHA {
-		inputPoolSize /= 2
+	// check sum of pool sizes against config db.connections-max
+	if readerPoolSize+writerPoolSize+maintPoolSize > cfgMaxConns {
+		return 0, 0, 0,
+			fmt.Errorf("reader-pool (size=%d) + writer-pool (size=%d) + maint-pool (size=%d) more than db.connections-max (%d). Increase the db.connections-max or decrease the pool-sizes", readerPoolSize, writerPoolSize, maintPoolSize, cfgMaxConns)
```
Comment (antekresic) [inline pkg/pgclient/config.go:228]: I'd use a switch case here, code would be easier to read and understand.

File: pkg/pgclient/config.go [hunk 3]
```diff
@@ -154,32 +158,76 @@ func (cfg *Config) GetConnectionStr() string {
...
 	}
-	return inputPoolSize, nil
+	return
```
Comment (antekresic) [inline pkg/pgclient/config.go:230]: Please no naked returns in such big functions.

File: pkg/pgclient/config.go [hunk 4]
```diff
@@ -154,32 +158,76 @@ func (cfg *Config) GetConnectionStr() string {
...
+
+	// calc reader pool size
+	readerFraction := defaultReaderFraction
+	if readOnly {
+		readerFraction = defaultConnFraction
+	}
+	readerPoolSize = calc(cfg.ReaderPoolSize, readerFraction)
+	if cfg.ReaderPoolSize == defaultPoolSize && readerPoolSize+writerPoolSize+maintPoolSize > cfgMaxConns {
+		readerPoolSize = readerPoolSize - maintPoolSize
+	}
```
Comment (antekresic) [inline pkg/pgclient/config.go:204]: I'd split this off in its own function that returns the pool sizes to contain the logic.

**Review by antekresic — CHANGES_REQUESTED** [review]

**Review by alejandrodnm — CHANGES_REQUESTED** [review]

File: pkg/pgclient/config.go [hunk 5]
```diff
@@ -154,32 +158,76 @@ func (cfg *Config) GetConnectionStr() string {
...
 }
 
-// GetPoolSize returns the max pool size based on the max_connections allowed by database and the defaultFraction.
-// Arg inputPoolSize is the pool size provided in the CLI flag.
-func (cfg *Config) GetPoolSize(poolName string, defaultFraction float64, inputPoolSize int) (int, error) {
-	maxConns, err := cfg.maxConn()
-	if err != nil {
-		return 0, fmt.Errorf("max connections: %w", err)
+// GetPoolSizes returns the pool sizes adjusted according to defaults, read only settings, and max connections available
+func (cfg *Config) GetPoolSizes(readOnly bool, dbMaxConns int) (readerPoolSize, writerPoolSize, maintPoolSize int, err error) {
```
Comment (alejandrodnm) [inline pkg/pgclient/config.go:162]: We could split this function in 2, getting the sizes and validating them. Maybe something like:

```
GetPoolSizes {
   poolSizes := doGetPoolSizes()
   err := validatePoolSizes(poolSizes)
}
```

WDYT?

File: pkg/pgclient/config.go [hunk 6]
```diff
@@ -154,32 +158,76 @@ func (cfg *Config) GetConnectionStr() string {
...
+	cfgMaxConns := cfg.MaxConnections
+	if cfgMaxConns == defaultMaxConns {
+		cfgMaxConns = int(float64(dbMaxConns) * 0.8)
```
[the remark this answers is not in the document]
Comment (jgpruitt) [reply pkg/pgclient/config.go:165]: agreed.

File: pkg/pgclient/config.go (hunk 3 above)
Comment (jgpruitt) [reply pkg/pgclient/config.go:230]: 👍

File: pkg/pgclient/config.go [hunk 7]
```diff
@@ -154,32 +158,76 @@ func (cfg *Config) GetConnectionStr() string {
...
+		)
 	}
 
-	switch {
-	case inputPoolSize == defaultPoolSize:
-		// For the default case, we need to take up defaultFraction of allowed connections.
-		poolSize := float64(maxConns) * defaultFraction
-		if cfg.UsesHA {
-			poolSize /= 2
+	calc := func(inputPoolSize int, fraction float64) int {
```
[the remark this answers is not in the document]
Comment (jgpruitt) [reply pkg/pgclient/config.go:174]: 👍

File: docs/configuration.md (hunk 1 above)
Comment (jgpruitt) [reply docs/configuration.md:77]: 👍

File: pkg/pgclient/config.go (hunk 5 above)
Comment (jgpruitt) [reply pkg/pgclient/config.go:162]: I'm splitting out the validation. I think once I do that and the other suggestions, the function is small enough to leave the calculation of the sizes inline.

File: pkg/pgclient/config.go (hunk 4 above)
Comment (jgpruitt) [reply pkg/pgclient/config.go]: i moved basically everything except this into other functions, so I think it's okay to leave it here. okay?

**Review by alejandrodnm — APPROVED** [review]

File: pkg/pgclient/config.go [hunk 8]
```diff
@@ -154,35 +159,81 @@ func (cfg *Config) GetConnectionStr() string {
 	return cfg.DbUri
 }
 
-// GetPoolSize returns the max pool size based on the max_connections allowed by database and the defaultFraction.
-// Arg inputPoolSize is the pool size provided in the CLI flag.
-func (cfg *Config) GetPoolSize(poolName string, defaultFraction float64, inputPoolSize int) (int, error) {
-	maxConns, err := cfg.maxConn()
+// GetPoolSizes returns the pool sizes adjusted according to defaults, read only settings, and max connections available
```
Comment (antekresic) [inline pkg/pgclient/config.go:162]: Nit:
```suggestion
// GetPoolSizes returns the pool sizes adjusted according to defaults, read only settings, and max connections available.
```

File: pkg/pgclient/config.go [hunk 9]
```diff
@@ -154,35 +159,81 @@ func (cfg *Config) GetConnectionStr() string {
...
+	case !readOnly && writerPoolSize < MinPoolSize:
+		return fmt.Errorf("writer pool size canot be less than %d: received %d", MinPoolSize, writerPoolSize)
+	case readerPoolSize < MinPoolSize:
+		return fmt.Errorf("reader pool size canot be less than %d: received %d", MinPoolSize, readerPoolSize)
+	case readerPoolSize+writerPoolSize+maintPoolSize > dbMaxConns:
+		return fmt.Errorf("reader-pool (size=%d) + writer-pool (size=%d) + maint-pool (size=%d) more than postgres max_connections (%d). Decrease the pool-sizes", readerPoolSize, writerPoolSize, maintPoolSize, dbMaxConns)
+	case readerPoolSize+writerPoolSize+maintPoolSize > cfgMaxConns:
+		return fmt.Errorf("reader-pool (size=%d) + writer-pool (size=%d) + maint-pool (size=%d) more than db.connections-max (%d). Increase the db.connections-max or decrease the pool-sizes", readerPoolSize, writerPoolSize, maintPoolSize, cfgMaxConns)
+	}
+	return nil
```
Comment (antekresic) [inline pkg/pgclient/config.go:222]: Minor nit: this is a minor semantic thing but putting the return nil into default case lets the reader know there is nothing after the switch. This forces you to do returns in every case of the switch statement (which is what we want here).

**Review by antekresic — APPROVED** [review]
🎉

Couple of minor nits which are no big deal.

File: pkg/pgclient/config.go (hunk 8 above)
Comment (jgpruitt) [reply pkg/pgclient/config.go]: 👍

File: pkg/pgclient/config.go (hunk 9 above)
Comment (jgpruitt) [reply pkg/pgclient/config.go]: 👍

**Pull request merged**